File: WalletKitCore/src/walletkit/WKWalletSweeper.c
//
//  WKWalletSweeper.c
//  WalletKitCore
//
//  Created by Ehsan Rezaie on 5/22/20
//  Copyright © 2020 Breadwinner AG. All rights reserved.
//
//  See the LICENSE file at the project root for license information.
//  See the CONTRIBUTORS file at the project root for a list of contributors.

#include "WKWalletP.h"
#include "WKWalletSweeperP.h"
#include "WKHandlersP.h"


extern WKWalletSweeper
wkWalletSweeperAllocAndInit (size_t sizeInBytes,
                                 WKNetworkType type,
                                 WKKey key,
                                 WKUnit unit) {
    assert (sizeInBytes >= sizeof (struct WKWalletSweeperRecord));
    assert (wkKeyHasSecret (key));
    
    WKWalletSweeper sweeper = calloc (1, sizeInBytes);
    
    sweeper->type = type;
    sweeper->handlers = wkHandlersLookup(type)->sweeper;
    sweeper->key = wkKeyTake (key);
    sweeper->unit = wkUnitTake (unit);
    
    return sweeper;
}

extern void
wkWalletSweeperRelease (WKWalletSweeper sweeper) {
    
    sweeper->handlers->release (sweeper);
    
    wkKeyGive (sweeper->key);
    wkUnitGive (sweeper->unit);
    
    memset (sweeper, 0, sizeof(*sweeper));
    free (sweeper);
}

extern WKWalletSweeperStatus
wkWalletSweeperAddTransactionFromBundle (WKWalletSweeper sweeper,
                                             OwnershipKept WKClientTransactionBundle bundle) {
    if (NULL == sweeper->handlers->addTranactionFromBundle) {
        assert(0);
        return WK_WALLET_SWEEPER_ILLEGAL_OPERATION;
    }
    
    return sweeper->handlers->addTranactionFromBundle (sweeper, bundle);
}

extern void
wkWalletManagerEstimateFeeBasisForWalletSweep (WKWalletSweeper sweeper,
                                                   WKWalletManager cwm,
                                                   WKWallet wallet,
                                                   WKCookie cookie,
                                                   WKNetworkFee fee) {
    WKFeeBasis feeBasis = sweeper->handlers->estimateFeeBasis (cwm,
                                                                     wallet,
                                                                     cookie,
                                                                     sweeper,
                                                                     fee);
    
    if (NULL != feeBasis)
        wkWalletGenerateEvent (wallet, wkWalletEventCreateFeeBasisEstimated (WK_SUCCESS, cookie, feeBasis));

    wkFeeBasisGive(feeBasis);
}

extern WKTransfer
wkWalletSweeperCreateTransferForWalletSweep (WKWalletSweeper sweeper,
                                                 WKWalletManager cwm,
                                                 WKWallet wallet,
                                                 WKFeeBasis estimatedFeeBasis) {
    WKTransfer transfer = sweeper->handlers->createTransfer (cwm, wallet, sweeper, estimatedFeeBasis);
    
    wkTransferGenerateEvent (transfer, (WKTransferEvent) {
        WK_TRANSFER_EVENT_CREATED
    });
    
    return transfer;
}

extern WKKey
wkWalletSweeperGetKey (WKWalletSweeper sweeper) {
    return wkKeyTake (sweeper->key);
}

extern WKAddress
wkWalletSweeperGetAddress (WKWalletSweeper sweeper) {
    return sweeper->handlers->getAddress (sweeper);
}

extern WKAmount
wkWalletSweeperGetBalance (WKWalletSweeper sweeper) {
    return sweeper->handlers->getBalance (sweeper);
}

extern WKWalletSweeperStatus
wkWalletSweeperValidate (WKWalletSweeper sweeper) {
    return sweeper->handlers->validate (sweeper);
}

File: WalletKitJava/WalletKitNative/src/main/java/com/blockset/walletkit/nativex/WKWalletSweeper.java
/*
 * Created by Michael Carrara <michael.carrara@breadwallet.com> on 7/1/19.
 * Copyright (c) 2019 Breadwinner AG.  All right reserved.
*
 * See the LICENSE file at the project root for license information.
 * See the CONTRIBUTORS file at the project root for a list of contributors.
 */
package com.blockset.walletkit.nativex;

import com.blockset.walletkit.nativex.library.WKNativeLibraryDirect;
import com.blockset.walletkit.nativex.utility.SizeT;
import com.google.common.base.Optional;
import com.sun.jna.Pointer;
import com.sun.jna.PointerType;

public class WKWalletSweeper extends PointerType {

    public WKWalletSweeper() {
        super();
    }

    public WKWalletSweeper(Pointer address) {
        super(address);
    }

    public static WKWalletSweeperStatus validateSupported(WKWalletManager cwm,
                                                          WKWallet wallet,
                                                          WKKey key) {
        return WKWalletSweeperStatus.fromCore(
                WKNativeLibraryDirect.wkWalletManagerWalletSweeperValidateSupported(
                        cwm.getPointer(),
                        wallet.getPointer(),
                        key.getPointer()
                )
        );
    }

    public static WKWalletSweeper createAsBtc(WKWalletManager cwm,
                                              WKWallet wallet,
                                              WKKey key) {
        return new WKWalletSweeper(
                WKNativeLibraryDirect.wkWalletManagerCreateWalletSweeper(
                        cwm.getPointer(),
                        wallet.getPointer(),
                        key.getPointer()
                )
        );
    }

    public WKKey getKey() {
        Pointer thisPtr = this.getPointer();

        return new WKKey(WKNativeLibraryDirect.wkWalletSweeperGetKey(thisPtr));
    }

    public Optional<WKAmount> getBalance() {
        Pointer thisPtr = this.getPointer();

        return Optional.fromNullable(
                WKNativeLibraryDirect.wkWalletSweeperGetBalance(
                        thisPtr
                )
        ).transform(WKAmount::new);
    }

    public Optional<WKAddress> getAddress() {
        Pointer thisPtr = this.getPointer();

        return Optional.fromNullable(
                WKNativeLibraryDirect.wkWalletSweeperGetAddress(thisPtr)
        ).transform(WKAddress::new);
    }

    public WKWalletSweeperStatus handleTransactionAsBtc(WKClientTransactionBundle bundle) {
        return WKWalletSweeperStatus.fromCore(
                WKNativeLibraryDirect.wkWalletSweeperAddTransactionFromBundle(
		      this.getPointer(),
		      bundle.getPointer())
        );
    }

    public WKWalletSweeperStatus validate() {
        Pointer thisPtr = this.getPointer();

        return WKWalletSweeperStatus.fromCore(
                WKNativeLibraryDirect.wkWalletSweeperValidate(thisPtr)
        );
    }

    public void give() {
        Pointer thisPtr = this.getPointer();

        WKNativeLibraryDirect.wkWalletSweeperRelease(thisPtr);
    }
}
